File: 2.kt
// Раздел 2. Задание 7.
fun searchCycle(str: String, clmn: Int): Boolean {
    // Функция Search_cycle - переборный цикл, работает он так: получает из main'а значение стержня,
    // и перебирает все цвета на условие, что кольцо каждого цвета находится на этом стержне
    var red = false
    var white = false
    var blue = false

    for (i in 0 until str.length - 1) {
        if (str[i] == 'R' && str[i + 1].digitToInt() == clmn) {
            red = true
        }
        if (str[i] == 'W' && str[i + 1].digitToInt() == clmn) {
            white = true
        }
        if (str[i] == 'B' && str[i + 1].digitToInt() == clmn) {
            blue = true
        }
        if (red && white && blue) {
            return true
        }
    }
    return false
}

fun main() {
    val output = mutableSetOf<Int>()
    println("Введите строку: R0W0B0W9W0R0B0")
    // Примеры строк:
    // R0R6B0W6W0W6B9
    // R0W0B0W9W0R0B0
    // B4
    // W0R3W7B0R2W9B5R1W4B8R6W0B3R9W2B7R5W1B4R8W6B2R0W3B9R1
    val rings = "R0W0B0W9W0R0B0"

    for (i in 0 until rings.length - 1) {
        if ((rings[i] == 'R' || rings[i] == 'W' || rings[i] == 'B') && rings[i + 1].isDigit()) {
            val column = rings[i + 1].digitToInt()
            if (searchCycle(rings, column)) {
                output.add(column)
            }
        }
    }

    if (output.isNotEmpty()) {
        println("Количество столбцов, на которых расположены и красные, и белые, и синие кольца: ${output.size}")
    } else {
        println("Нету столбцов с кольцами всех трёх цветов")
    }
}

File: 2.py
# Раздел 2. Задание 7.

def search_cycle(s: str, clmn: int) -> int:

    red = white = blue = False
    
    for j in range(len(s) - 1):
        if s[j] == 'R' and int(s[j+1]) == clmn:
            red = True
        if s[j] == 'W' and int(s[j+1]) == clmn:
            white = True
        if s[j] == 'B' and int(s[j+1]) == clmn:
            blue = True
        if red and white and blue:
            return clmn
    return -1

str1 = input("Введите строку: ").strip()
# Примеры строк:
# R0R6B0W6W0W6B9
# R0W0B0W9W0R0B0
# B4
# W0R3W7B0R2W9B5R1W4B8R6W0B3R9W2B7R5W1B4R8W6B2R0W3B9R1

# Обработка строки
output = set()

for i in range(len(str1) - 1):
    char, next_char = str1[i], str1[i+1]
    if char in {'R', 'W', 'B'} and next_char.isdigit():
        column = int(next_char)
        if search_cycle(str1, column) != -1:
            output.add(column)

# Вывод результата
if output:
    print(f"Количество столбцов с кольцами всех цветов: {len(output)}")
else:
    print("Нет столбцов с кольцами всех трех цветов")
